ui/job-view/Push.jsx

```diff
@@ -46,48 +46,52 @@ export default class Push extends React.Component {
 
-    if (Notification.permission !== 'granted' || watched === 'none') {
-      return;
-    }
-
-    const nextCounts = nextProps.push.job_counts;
-    if (last_job_counts) {
-      const nextUncompleted = nextCounts.pending + nextCounts.running;
-      const lastUncompleted = last_job_counts.pending + last_job_counts.running;
-
-      const nextCompleted = nextCounts.completed;
-      const lastCompleted = last_job_counts.completed;
-
-      let message;
-      if (lastUncompleted > 0 && nextUncompleted === 0) {
-        message = 'Push completed';
-        this.setState({ watched: 'none' });
-      } else if (watched === 'job' && lastCompleted < nextCompleted) {
-        const completeCount = nextCompleted - lastCompleted;
-        message = completeCount + ' jobs completed';
+    try {
+      if (Notification.permission !== 'granted' || watched === 'none') {
+        return;
       }
 
-      if (message) {
-        const notification = new Notification(message, {
-          body: `${repoName} rev ${revision.substring(0, 12)}`,
-          tag: pushId,
-        });
-
-        notification.onerror = (event) => {
-          this.thNotify.send(`${event.target.title}: ${event.target.body}`, 'danger');
-        };
-
-        notification.onclick = (event) => {
-          if (this.container) {
-            this.container.scrollIntoView();
-            event.target.close();
-          }
-        };
+      const nextCounts = nextProps.push.job_counts;
+      if (last_job_counts) {
+        const nextUncompleted = nextCounts.pending + nextCounts.running;
+        const lastUncompleted = last_job_counts.pending + last_job_counts.running;
+
+        const nextCompleted = nextCounts.completed;
+        const lastCompleted = last_job_counts.completed;
+
+        let message;
+        if (lastUncompleted > 0 && nextUncompleted === 0) {
+          message = 'Push completed';
+          this.setState({ watched: 'none' });
+        } else if (watched === 'job' && lastCompleted < nextCompleted) {
+          const completeCount = nextCompleted - lastCompleted;
+          message = completeCount + ' jobs completed';
+        }
+
+        if (message) {
+          const notification = new Notification(message, {
+            body: `${repoName} rev ${revision.substring(0, 12)}`,
+            tag: pushId,
+          });
+
+          notification.onerror = (event) => {
+            this.thNotify.send(`${event.target.title}: ${event.target.body}`, 'danger');
+          };
+
+          notification.onclick = (event) => {
+            if (this.container) {
+              this.container.scrollIntoView();
+              event.target.close();
+            }
+          };
+        }
       }
-    }
 
-    if (nextCounts) {
-      this.setState({ last_job_counts: Object.assign({}, nextCounts) });
+      if (nextCounts) {
+        this.setState({ last_job_counts: Object.assign({}, nextCounts) });
+      }
+    } catch (e) {
```

Comment:
This try-catch seems a bit dangerous - since it will catch any exception in this block and silently hide it, regardless of whether it has anything to do with the Notifications API being available. I think it would be best to just use feature detection (ie `watched === 'none' || !('Notification' in window && Notification.permission === 'granted')`) and also perhaps to move the duplicated checks to another file? (eg `helpers/notifications.js`?)